File: backend/application/use_cases/retry_job.py
from domain.entities import JobMetadata, JobStatus
from domain.errors import InvalidJobStateError, JobNotFoundError
from domain.interfaces import JobRepository


class RetryJobUseCase:
    """Reset a failed job for reprocessing."""

    def __init__(self, repo: JobRepository) -> None:
        self._repo = repo

    def execute(self, job_id: str, resume: bool = False) -> JobMetadata:
        """Reset a failed job for reprocessing.

        If resume=True and last_offset_ms exists, keeps extracted audio
        and retries transcription from the last known offset.
        Otherwise, performs a full retry from scratch.
        """
        current = self._repo.get(job_id)
        if current is None:
            raise JobNotFoundError(job_id)

        if current.status != JobStatus.FAILED:
            raise InvalidJobStateError(
                job_id,
                current.status.value,
                [JobStatus.FAILED.value],
            )

        if resume and current.last_offset_ms is not None:
            return self._repo.update(
                job_id,
                status=JobStatus.PENDING,
                error=None,
                progress=0.20,
                extraction_progress=1.0,
                transcription_progress=0.0,
                formatting_progress=0.0,
                started_at=None,
                completed_at=None,
                duration_seconds=None,
            )

        return self._repo.update(
            job_id,
            status=JobStatus.PENDING,
            progress=0.0,
            error=None,
            extraction_progress=0.0,
            transcription_progress=0.0,
            formatting_progress=0.0,
            started_at=None,
            completed_at=None,
            duration_seconds=None,
            last_offset_ms=None,
        )

File: src/main/application/use-cases/retry-job.ts
import type { JobMetadata } from '../../../shared/types'
import { JobStatus } from '../../../shared/types'
import { InvalidJobStateError, JobNotFoundError } from '../../domain/errors'
import type { DrizzleJobRepository } from '../../infrastructure/repositories/drizzle-job-repository'

export class RetryJobUseCase {
  constructor(private readonly repo: DrizzleJobRepository) {}

  execute(jobId: string, resume = false): JobMetadata {
    const current = this.repo.get(jobId)
    if (current === null) throw new JobNotFoundError(jobId)

    if (current.status !== JobStatus.FAILED) {
      throw new InvalidJobStateError(jobId, current.status, [JobStatus.FAILED])
    }

    if (resume && current.lastOffsetMs !== null) {
      // Keep extracted audio; resume transcription from last offset
      return this.repo.update(jobId, {
        status: JobStatus.PENDING,
        error: null,
        progress: 0.2,
        extractionProgress: 1.0,
        transcriptionProgress: 0,
        formattingProgress: 0,
        startedAt: null,
        completedAt: null,
        durationSeconds: null,
      })
    }

    // Full retry from scratch
    return this.repo.update(jobId, {
      status: JobStatus.PENDING,
      progress: 0,
      error: null,
      extractionProgress: 0,
      transcriptionProgress: 0,
      formattingProgress: 0,
      startedAt: null,
      completedAt: null,
      durationSeconds: null,
      lastOffsetMs: null,
    })
  }
}
